File: requests-api/requests_app/domain/services/correction_targets.py
"""Normalização e catálogo de campos marcáveis na devolução."""

from __future__ import annotations

import re
from typing import Any

_KEY_RE = re.compile(r"^[a-z][a-z0-9_]{0,63}$")
_MAX_TARGETS = 32

# Catálogo canônico por type_code (labels PT só para docs/API; UI MFE espelha).
CORRECTION_TARGET_CATALOG: dict[str, tuple[str, ...]] = {
    "invoice-issuance": (
        "recipient",
        "invoice_type",
        "items",
        "freight",
        "extras",
    ),
    "raw-material-creation": (
        "description",
        "unit",
        "notes",
    ),
}


def allowed_correction_targets(type_code: str) -> frozenset[str]:
    return frozenset(CORRECTION_TARGET_CATALOG.get(type_code) or ())


def normalize_correction_targets(
    raw: Any,
    *,
    type_code: str | None = None,
) -> list[str]:
    """Aceita list/tuple; filtra inválidos; se type_code, restringe ao catálogo."""
    if raw is None:
        return []
    if isinstance(raw, str):
        candidates = [raw]
    elif isinstance(raw, (list, tuple)):
        candidates = list(raw)
    else:
        return []

    allowed = allowed_correction_targets(type_code) if type_code else None
    out: list[str] = []
    seen: set[str] = set()
    for item in candidates:
        if not isinstance(item, str):
            continue
        key = item.strip().lower()
        if not key or key in seen:
            continue
        if not _KEY_RE.match(key):
            continue
        if allowed is not None and key not in allowed:
            continue
        seen.add(key)
        out.append(key)
        if len(out) >= _MAX_TARGETS:
            break
    return out

File: plugins/my-requests/src/content/correctionTargets.ts
/** Catálogo de campos/seções marcáveis na devolução — espelha requests-api. */

export type CorrectionTargetOption = {
  id: string;
  label: string;
  /** Wizard step id quando aplicável (invoice). */
  wizardStepId?: string;
};

const INVOICE_TARGETS: CorrectionTargetOption[] = [
  { id: "recipient", label: "Destinatário", wizardStepId: "recipient" },
  { id: "invoice_type", label: "Tipo de NF", wizardStepId: "invoiceType" },
  { id: "items", label: "Itens", wizardStepId: "items" },
  { id: "freight", label: "Transporte", wizardStepId: "freight" },
  { id: "extras", label: "Adicionais (peso, volumes, observação)", wizardStepId: "extras" },
];

const RAW_MATERIAL_TARGETS: CorrectionTargetOption[] = [
  { id: "description", label: "Descrição" },
  { id: "unit", label: "Unidade" },
  { id: "notes", label: "Observações" },
];

const BY_TYPE: Record<string, CorrectionTargetOption[]> = {
  "invoice-issuance": INVOICE_TARGETS,
  "raw-material-creation": RAW_MATERIAL_TARGETS,
};

export function correctionTargetOptionsForType(
  typeCode: string | null | undefined,
): CorrectionTargetOption[] {
  if (!typeCode) return [];
  return BY_TYPE[typeCode] ?? [];
}

export function correctionTargetLabel(
  typeCode: string | null | undefined,
  targetId: string,
): string {
  const hit = correctionTargetOptionsForType(typeCode).find((row) => row.id === targetId);
  return hit?.label ?? targetId;
}

export function correctionTargetLabels(
  typeCode: string | null | undefined,
  targetIds: string[] | null | undefined,
): string[] {
  if (!targetIds?.length) return [];
  return targetIds.map((id) => correctionTargetLabel(typeCode, id));
}

/** Map API correction target ids → wizard step ids (invoice). */
export function wizardStepIdsForCorrectionTargets(
  targetIds: string[] | null | undefined,
): string[] {
  if (!targetIds?.length) return [];
  const map = new Map(
    INVOICE_TARGETS.filter((row) => row.wizardStepId).map((row) => [
      row.id,
      row.wizardStepId as string,
    ]),
  );
  const out: string[] = [];
  const seen = new Set<string>();
  for (const id of targetIds) {
    const step = map.get(id);
    if (!step || seen.has(step)) continue;
    seen.add(step);
    out.push(step);
  }
  return out;
}
